desktop-widgets/btdeviceselectiondialog.cpp

```diff
@@ -153,6 +155,66 @@ void BtDeviceSelectionDialog::on_changeDeviceState_clicked()
 #endif
 }
 
+static quint32 deviceInfoToClass(const QBluetoothDeviceInfo &info)
+{
+	return (info.minorDeviceClass() << 2) |
+	       (info.majorDeviceClass() << 8) |
+	       ((quint32)info.serviceClasses() << 13);
+}
+
+static QString addressOrUuid(const QBluetoothDeviceInfo &info)
+{
+	return info.address().isNull() ? info.deviceUuid().toString() : info.address().toString();
+}
+
+// Compare Bluetooth devices by address or UUID.
+// When using QBluetoothDeviceInfo::operator==(), scanned and loaded (from the settings)
+// device-infos don't compare as equal, for yet unknown reasons.
+static bool sameDevice(const QBluetoothDeviceInfo &d1, const QBluetoothDeviceInfo &d2)
+{
+	return addressOrUuid(d1) == addressOrUuid(d2);
+}
+
+void BtDeviceSelectionDialog::saveDeviceList()
+{
+	int numRows = ui->discoveredDevicesList->count();
+	QSettings s;
+	s.beginGroup("Bluetooth");
+	s.beginWriteArray("devices", numRows);
+	for (int i = 0; i < numRows; ++i) {
+		QBluetoothDeviceInfo info = ui->discoveredDevicesList->item(i)->data(Qt::UserRole).value<QBluetoothDeviceInfo>();
+		s.setArrayIndex(i);
+		s.setValue("name", info.name());
+		s.setValue("address", addressOrUuid(info));
+		s.setValue("class", deviceInfoToClass(info));
+	}
+	s.endArray();
+	s.endGroup();
```

Comment:
this of course is outside the Pref structure. But yeah, ok, whatever.
I do however wonder if this is still the right direction to follow. And part of this will require more staring at the code. Because I don't think it makes any sense at all to save all of the discovered devices. If I scan here in my office I see more than 40 BLE devices... and not a single one of them is a dive computer and therefore worth remembering.
I think that the (up to) four remembered dive computers with their connections are a much more useful thing to remember.
What am I missing?

Reply:
> this of course is outside the Pref structure. 

Haven't we decided to keep things like `lastUsedPath` outside of prefs because it is *such* a pain 
 for dubious gain to add new fields?

>  If I scan here in my office I see more than 40 BLE devices

Wow. I see two in my office [laptop + phone]. :)

To understand the save-the-known-devices, you have to look at the history of this PR. This was the only way to make all three desktop platforms behave the same. On Linux you get the cached devices immediately but then it scans. On windows you have to decide whether you want cached or scan and scanning is not asynchronous, etc.

The idea is that the user only very rarely uses this dialog. If they select dive computer XYZ, then libdivecomputer tells us if this is a BT device and in this case we can use simple heuristics to choose the correct BT device. Only if they have more than one XYZ dive-computers they should even need to enter this dialog. For non-BT computers, this dialog will be disabled. At least that's my vision. :)

Reply:
So... the whole prefs thing needs to be cleaned up. I was too frustrated and too distracted last week, but I want to push further into this and turn it into something that I don't hate.
Your read only work is IMHO part of that.
My comment above was more like an "eye roll". All the work we did to only read and write settings in one spot. Except when we aren't

Reply:
Oh, and I checked again. 47 BLE devices